lib/UserAgentApplication.ts

```diff
@@ -445,221 +728,330 @@ namespace MSAL {
+                } else if (requestInfo.requestType === Constants.login) {
+                    tokenReceivedCallback = this._tokenReceivedCallback;
                     token = requestInfo.parameters[Constants.idToken];
+                    tokenType = Constants.idToken;
                 }
 
                 try {
-                    if (callback)
-                        callback(this._cacheStorage.getItem(Constants.errorDescription), token, this._cacheStorage.getItem(Constants.error));
+                    if (tokenReceivedCallback) {
+                        tokenReceivedCallback(this._cacheStorage.getItem(Constants.errorDescription), token, this._cacheStorage.getItem(Constants.error), tokenType);
+                    }
                 } catch (err) {
-                    this._requestContext.logger.error('Error occurred in user defined callback function: ' + err)
+                    this._requestContext.logger.error('Error occurred in token received callback function: ' + err);
                 }
 
                 if (this._interactionMode !== this._interactionModes.popUp) {
-                    window.location.hash = '';
-                    if (this.navigateToLoginRequestUrl && window.location.href.replace('#', '') !== this._cacheStorage.getItem(Constants.loginRequest))
+                    window.location.hash = "";
+                    if (this.navigateToLoginRequestUrl && window.location.href.replace("#", "") !== this._cacheStorage.getItem(Constants.loginRequest))
                         window.location.href = this._cacheStorage.getItem(Constants.loginRequest);
                 }
             }
         }
 
-        private saveAccessToken(requestInfo: RequestInfo): void {
-            if (requestInfo.parameters.hasOwnProperty('scope')) {
-                this.user = this.getUser();
-                let scopes = requestInfo.parameters['scope'];
-                let consentedScopes = scopes.split(' ');
-                let accessTokenCacheItems: Array<AccessTokenCacheItem> = this._cacheStorage.getAllAccessTokens(this.clientId, this.authority);
+        /**
+         * This method must be called for processing the response received from AAD. It extracts the hash, processes the token or error, saves it in the cache and calls the registered callbacks with the result.
+         * @param {RequestInfo} requestInfo an object created from the redirect response from AAD comprising of the keys - parameters, requestType, stateMatch, stateResponse and valid.
+         * @param {User} user user object for which scopes are consented for. The default user is the logged in user.
+         * @param {ClientInfo} clientInfo clientInfo received as part of the response comprising of fields uid and utid.
+         * @param {IdToken} idToken idToken received as part of the response.
+         * @ignore
+         */
+        private saveAccessToken(requestInfo: RequestInfo, user: User, clientInfo: string, idToken: IdToken): void {
```

Comment:
We get the decipher the tenant id from the authority url. Are you sure it needs to be updated this way? @kpanwar